lib/Coroutine.php

```diff
@@ -15,126 +15,150 @@ final class Coroutine implements Promise
+        $immediate = true;
+
+        /** @var \Throwable|null Promise failure reason when executing next coroutine step, null at all other times. */
+        $exception = null;
+
+        /** @var mixed Promise success value when executing next coroutine step, null at all other times. */
+        $value = null;
+
         /**
-         * @param \Throwable|null $exception Exception to be thrown into the generator.
-         * @param mixed           $value Value to be sent into the generator.
+         * @param \Throwable|null $e Exception to be thrown into the generator.
+         * @param mixed           $v Value to be sent into the generator.
          */
-        $this->onResolve = function ($exception, $value) {
-            $this->exception = $exception;
-            $this->value = $value;
+        $onResolve = function ($e, $v) use ($generator, &$exception, &$value, &$immediate, &$onResolve) {
+            $exception = $e;
+            $value = $v;
 
-            if (!$this->immediate) {
-                $this->immediate = true;
+            if (!$immediate) {
+                $immediate = true;
                 return;
             }
 
             try {
                 do {
-                    if ($this->exception) {
+                    if ($exception) {
                         // Throw exception at current execution point.
-                        $yielded = $this->generator->throw($this->exception);
+                        $yielded = $generator->throw($exception);
                     } else {
                         // Send the new value and execute to next yield statement.
-                        $yielded = $this->generator->send($this->value);
+                        $yielded = $generator->send($value);
                     }
 
                     if (!$yielded instanceof Promise) {
-                        if (!$this->generator->valid()) {
-                            $this->resolve($this->generator->getReturn());
-                            $this->onResolve = null;
+                        if (!$generator->valid()) {
+                            $this->resolve($generator->getReturn());
+                            $onResolve = null;
                             return;
                         }
 
-                        $yielded = $this->transform($yielded);
+                        $yielded = self::transform($yielded, $generator);
                     }
 
-                    $this->immediate = false;
-                    $yielded->onResolve($this->onResolve);
-                } while ($this->immediate);
+                    $immediate = false;
+                    $yielded->onResolve($onResolve);
+                } while ($immediate);
 
-                $this->immediate = true;
+                $immediate = true;
             } catch (\Throwable $exception) {
-                $this->fail($exception);
-                $this->onResolve = null;
+                try {
```

Comment:
Can we wrap the `try` started on line 97 in a single try/catch rather than separate try/catch blocks in catch and finally?

Reply:
Done.